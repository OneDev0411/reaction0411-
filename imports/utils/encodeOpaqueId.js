import { curry } from "ramda";

/**
 * @name encodeOpaqueId
 * @method
 * @memberof GraphQL/Transforms
 * @summary Transforms an internal ID to an opaque ID
 * @param {String} namespace The namespace of the ID
 * @param {String} id The ID to transform
 * @returns {String} An opaque ID
 */
const encodeOpaqueId = curry((namespace, id) => {
  if (typeof id !== "string" && typeof id !== "number") return id;
  const unencoded = `${namespace}:${id}`;
  return Buffer.from(unencoded).toString("base64");
});

export default encodeOpaqueId;
