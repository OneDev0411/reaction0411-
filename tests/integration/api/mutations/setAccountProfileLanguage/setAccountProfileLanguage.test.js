import importAsString from "@reactioncommerce/api-utils/importAsString.js";
import encodeOpaqueId from "@reactioncommerce/api-utils/encodeOpaqueId.js";
import Factory from "/tests/util/factory.js";
import TestApp from "/tests/util/TestApp.js";

const SetAccountProfileLanguageMutation = importAsString("./SetAccountProfileLanguageMutation.graphql");

jest.setTimeout(300000);

let testApp;
let setAccountProfileLanguage;
let shopId;
let mockUserAccount;
let accountOpaqueId;

beforeAll(async () => {
  testApp = new TestApp();
  await testApp.start();
  shopId = await testApp.insertPrimaryShop({
    languages: [
      { label: "mockLabel", i18n: "mockI18n", enabled: false },
      { label: "English", i18n: "EN", enabled: true }
    ]
  });
  setAccountProfileLanguage = testApp.mutate(SetAccountProfileLanguageMutation);

  const adminGroup = Factory.Group.makeOne({
    _id: "adminGroup",
    createdBy: null,
    name: "admin",
    permissions: ["reaction:legacy:accounts/update:language"],
    slug: "admin"
  });
  await testApp.collections.Groups.insertOne(adminGroup);

  mockUserAccount = Factory.Account.makeOne({
    _id: "mockUserId",
    groups: [adminGroup._id]
  });

  accountOpaqueId = encodeOpaqueId("reaction/account", mockUserAccount._id);

  await testApp.createUserAndAccount(mockUserAccount, ["reaction:legacy:accounts/update:language"]);
});

// There is no need to delete any test data from collections because
// testApp.stop() will drop the entire test database. Each integration
// test file gets its own test database.
afterAll(() => testApp.stop());

afterEach(async () => {
  await testApp.clearLoggedInUser();
});

test("user can set account profile language", async () => {
  await testApp.setLoggedInUser(mockUserAccount);

  let result;
  try {
    result = await setAccountProfileLanguage({
      input: { accountId: accountOpaqueId, language: "EN" }
    });
  } catch (error) {
    expect(error).toBeUndefined();
    return;
  }
  expect(result.setAccountProfileLanguage.account.language).toEqual("EN");
});

test("user cannot set account profile language if its not enabled", async () => {
  await testApp.setLoggedInUser(mockUserAccount);

  try {
    await setAccountProfileLanguage({
      input: { accountId: accountOpaqueId, language: "mockI18n" }
    });
  } catch (error) {
    expect(error).toMatchSnapshot();
  }
});

test("user can not set account profile language if not logged in", async () => {
  try {
    await setAccountProfileLanguage({
      input: { accountId: accountOpaqueId, language: "INR" }
    });
  } catch (error) {
    expect(error).toMatchSnapshot();
  }
});
