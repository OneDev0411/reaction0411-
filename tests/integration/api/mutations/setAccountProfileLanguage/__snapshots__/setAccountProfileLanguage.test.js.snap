// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user can not set account profile language if not logged in 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "FORBIDDEN",
      "exception": Object {
        "details": Object {},
        "error": "access-denied",
        "eventData": Object {},
        "isClientSafe": true,
        "reason": "Access Denied",
      },
    },
    "locations": Array [
      Object {
        "column": 3,
        "line": 2,
      },
    ],
    "message": "Access Denied",
    "path": Array [
      "setAccountProfileLanguage",
    ],
  },
]
`;

exports[`user cannot set account profile language if its not enabled 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "INTERNAL_SERVER_ERROR",
      "exception": Object {
        "details": Object {},
        "error": "invalid-argument",
        "eventData": Object {},
        "isClientSafe": true,
        "reason": "Shop does not define any enabled language with code \\"mockI18n\\"",
      },
    },
    "locations": Array [
      Object {
        "column": 3,
        "line": 2,
      },
    ],
    "message": "Shop does not define any enabled language with code \\"mockI18n\\"",
    "path": Array [
      "setAccountProfileLanguage",
    ],
  },
]
`;
